zipline/pipeline/loaders/quarter_estimates.py

```diff
@@ -0,0 +1,287 @@
+    # When we get 0, this actually means we're in Q1 of the previous year,
+    # so we need to adjust.
+    to_adjust = result_qtrs[result_qtrs == 0].index
+    result_years.iloc[to_adjust] -= 1
+    result_qtrs.iloc[to_adjust] = 4
+    return result_years, result_qtrs
+
+
+def calc_backward_shift(yrs, qtrs, num_qtrs_shift):
+    """
+    Calculate the new years and quarters based shifting the specified number
+    of quarters backwards.
+
+    Parameters
+    ----------
+    yrs : np.Series
+        The starting years.
+    qtrs : np.Series
+        The starting quarters.
+    num_qtrs_shift : int
+        The number of quarters to shift backward.
+
+
+    Returns
+    -------
+    result_years : pd.Series
+        A series contains the new years.
+    result_qtrs : pd.Series
+        A series that contains the new quarters.
+    """
+
+    if num_qtrs_shift < 0:
+        raise AssertionError("Must pass a number of quarters >= 0")
+    result_qtrs = 4 - (num_qtrs_shift - qtrs) % 4
+    # Subtract 1 year since we go backwards at least `qtrs` number of quarters.
+    result_years = yrs - (num_qtrs_shift - qtrs) // 4 - 1
+    # Find cases where we aren't shifting enough quarters backwards to cross
+    # a year boundary and correct for these.
+    no_yr_boundary_crossed = qtrs[qtrs > num_qtrs_shift].index
+    # Set the year back to the original.
+    result_years.iloc[no_yr_boundary_crossed] = yrs.iloc[
+        no_yr_boundary_crossed
+    ]
+    result_qtrs.iloc[no_yr_boundary_crossed] = qtrs.iloc[
+                                                   no_yr_boundary_crossed
+                                               ] - num_qtrs_shift
+    return result_years, result_qtrs
+
+
+def required_estimates_fields(columns):
+    """
+    Compute the set of resource columns required to serve
+    `columns`.
+    """
+    # These metadata columns are used to align event indexers.
+    return {
+        TS_FIELD_NAME,
+        SID_FIELD_NAME,
+        EVENT_DATE_FIELD_NAME,
+        FISCAL_QUARTER_FIELD_NAME,
+        FISCAL_YEAR_FIELD_NAME
+    }.union(
+        # We also expect any of the field names that our loadable columns
+        # are mapped to.
+        viewvalues(columns),
+    )
+
+
+def validate_column_specs(events, columns):
+    """
+    Verify that the columns of ``events`` can be used by a
+    QuarterEstimatesLoader to serve the BoundColumns described by
+    `columns`.
+    """
+    required = required_estimates_fields(columns)
+    received = set(events.columns)
+    missing = required - received
+    if missing:
+        raise ValueError(
+            "QuarterEstimatesLoader missing required columns {missing}.\n"
+            "Got Columns: {received}\n"
+            "Expected Columns: {required}".format(
+                missing=sorted(missing),
+                received=sorted(received),
+                required=sorted(required),
+            )
+        )
+
+
+def cross_product(df1, df2):
+    df1['key'] = 1
+    df2['key'] = 1
```

Comment:
`key` seems a little too likely to collide with existing keys. Could we use a uuid4 or something? Also, could we move this to pandas_utils?